tests/test-dladdr.c

```diff
@@ -0,0 +1,176 @@
+}
+
+/**
+ * @brief check information returned by dladdr
+ * @param addr address to check
+ * @param addrsym
+ * @param expected check against expected result
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr( void *addr, char *addrsym, ExpectedResult expected_result )
+{
+    Dl_info info;
+    int result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match)
+                 || (expected_result == PassWithDifferentAddress && sym_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+    }
+    printf( "check address %p which has symbol '%s' -> %s%s", addr, addrsym, passed ? "passed" : "failed", verbose || !passed ? " " : "\n" );
+    if( verbose || !passed )
+    {
+        if( !result )
+            printf( "(could not get symbol information for address %p)\n", addr );
+        else
+            print_dl_info( &info, "(", ")\n");
+    }
+    return !passed;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+#ifdef _MSC_VER
+#include <windows.h>
+#else
+/* link to import thunk */
+HMODULE __stdcall GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T __stdcall VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+#endif
```

Comment:
Reason for those failing test and only under MSVC is simple: First you use preprocessor to replace keyword HMODULE by void* and then only under MSVC include header file windows.h which is typdefing keyword HMODULE (now replaed to void*) to something else. This would result in compile errors. As windows.h is not included under MinGW there is no compile error.

This code is really suspicous. There is nothing compiler specific, test should work with any compiler which targets win32 and therefore there should not be any #ifdef compiler parts.

Reply:
> This code is really suspicous. 

There are basically two ways to use functions from a DLL:

1. via an import thunk, which is the case if no dllimport is used.
2. via a direct access to the IAT using dllimport.

Which of these methods is used is a hidden implementation detail of the header where the corresponding function is declared and the code used forces the function in question to be included via path 1.

> There is nothing compiler specific, test should work with any compiler which targets win32 and therefore there should not be any #ifdef compiler parts.

Forcing the mentioned function to use path 1 was implemented for and test with mingw, but does not work with msvc.

Reply:
> but does not work with msvc.

Ok, in this case enable those tests via path 1 only when not using msvc.